src/Infrastructure/DoctrinePostPersistSubscriberCreateAddressChange.php

```diff
@@ -0,0 +1,127 @@
+
+	private const IS_PRIVATE_PERSON = 0;
+	private const IS_COMPANY = 1;
+	private $myEntityManager;
+
+	public function __construct( EntityManager $entityManager ) {
+		$this->myEntityManager = $entityManager;
+	}
+
+	public function getSubscribedEvents(): array {
+		return [ Events::postPersist ];
+	}
+
+	public function postPersist( LifecycleEventArgs $args ): void {
+		$entity = $args->getObject();
+
+		if ( !$entity instanceof Donation && !$entity instanceof Application ) {
+			return;
+		}
+
+		if ( $entity instanceof Donation && $entity->getDonor() === null ) {
+			return;
+		}
+
+		$addressOptions = $this->getAddressOptions( $entity );
+
+		if ( $addressOptions['address'] == null ) {
+			return;
+		}
+
+		$addressChangeBuilder = AddressChangeBuilder::create( null, $addressOptions['address'] );
+
+		if ( $entity instanceof Donation ) {
+			$addressChangeBuilder->forDonation( $entity->getId() );
+		} elseif ( $entity instanceof Application ) {
+			$addressChangeBuilder->forMembership( $entity->getId() );
+		}
+
+		if ( $addressOptions['addressType'] === self::IS_PRIVATE_PERSON ) {
+			$addressChangeBuilder->forPerson();
+		} elseif ( $addressOptions['addressType'] === self::IS_COMPANY ) {
+			$addressChangeBuilder->forCompany();
+		}
+
+		$addressChange = $addressChangeBuilder->build();
+
+		$this->myEntityManager->persist( $addressChange );
+		$this->myEntityManager->flush();
+	}
+
+	private function getAddressOptions( object $entity ): array {
+		if ( $entity instanceof Donation ) {
+			$donor = $entity->getDonor();
+			$name = $donor->getName();
+			$address = $donor->getPhysicalAddress();
+
+			return [
+				'address' => $this->buildAddress( $name, $address ),
+				'addressType' => $this->getAddressType( $name )
+			];
+
+		} elseif ( $entity instanceof Application ) {
+			$applicant = $entity->getApplicant();
+			$name = $applicant->getName();
+			$address = $applicant->getPhysicalAddress();
+
+			return [
+				'address' => $this->buildAddress( $name, $address ),
+				'addressType' => $this->getAddressType( $name )
+			];
+
+		}
+
+		return [ 'address' => null, 'addressType' => null ];
+	}
+
+	private function buildAddress( object $name, object $address ): Address {
+		switch ( $this->getAddressType( $name ) ) {
+			case self::IS_PRIVATE_PERSON:
+				return Address::newPersonalAddress(
+					$name->getSalutation(),
+					$name->getTitle(),
+					$name->getFirstName(),
+					$name->getLastName(),
+					$address->getStreetAddress(),
+					$address->getPostalCode(),
+					$address->getCity(),
+					$address->getCountryCode()
+				);
+			case self::IS_COMPANY:
+				return Address::newCompanyAddress(
+					$name->getCompanyName(),
+					$address->getStreetAddress(),
+					$address->getPostalCode(),
+					$address->getCity(),
+					$address->getCountryCode()
+				);
+			default:
+				return null;
+		}
+	}
+
+	private function getAddressType( object $name ): int {
```

Comment:
In cases like these where you have a [Union Typehint](https://stackoverflow.com/questions/46775489/is-it-possible-to-type-hint-more-than-one-type) with types that have the same "interface" (meaning they share the same method, not that they share the same interface) it's better to have no PHP typehint and specify the union type in a doc comment. PHP 8 will fix this.

Also, I'd rather have `AddressChangedBuilder` passed in here and called directly instead of "indirecting" with constants. You'll decrease the overall cyclomatic complexity of this class.